sway-core/src/semantic_analysis/ast_node/expression/typed_expression.rs

```diff
@@ -62,21 +62,47 @@ impl ty::TyExpression {
             span: call_path.span(),
         };
         let arguments = VecDeque::from(arguments);
-        let method = check!(
+        let decl_id = check!(
             resolve_method_name(ctx, &method_name_binding, arguments.clone()),
             return err(warnings, errors),
             warnings,
             errors
         );
-        instantiate_function_application_simple(
-            call_path,
-            HashMap::new(),
-            arguments,
-            method,
-            None,
-            None,
+        let method = check!(
+            CompileResult::from(de_get_function(
+                decl_id.clone(),
+                &method_name_binding.span()
+            )),
+            return err(warnings, errors),
+            warnings,
+            errors
+        );
+        // check that the number of parameters and the number of the arguments is the same
+        check!(
+            check_function_arguments_arity(arguments.len(), &method, &call_path),
+            return err(warnings, errors),
+            warnings,
+            errors
+        );
+        let args_and_names = method
+            .parameters
+            .iter()
+            .zip(arguments.into_iter())
+            .map(|(param, arg)| (param.name.clone(), arg))
+            .collect::<Vec<(_, _)>>();
+        let exp = ty::TyExpression {
+            expression: ty::TyExpressionVariant::FunctionApplication {
+                call_path,
+                contract_call_params: HashMap::new(),
+                arguments: args_and_names,
+                function_decl_id: decl_id,
+                self_state_idx: None,
+                selector: None,
+            },
+            return_type: method.return_type,
```

Comment:
Not too important, but we could avoid cloning param names by taking the ret type first, since it is a smaller sized type:
```suggestion
        let return_type = method.return_type.clone();
        let args_and_names = method
            .parameters
            .into_iter()
            .zip(arguments.into_iter())
            .map(|(param, arg)| (param.name, arg))
            .collect::<Vec<(_, _)>>();
        let exp = ty::TyExpression {
            expression: ty::TyExpressionVariant::FunctionApplication {
                call_path,
                contract_call_params: HashMap::new(),
                arguments: args_and_names,
                function_decl_id: decl_id,
                self_state_idx: None,
                selector: None,
            },
            return_type,
```